### PR TITLE
[css] Add tests for `justify-self` on block-level elements

### PR DESCRIPTION
**Pull request by davidtranhq**
As per https://www.w3.org/TR/css-align-3/#justify-self-property , the `justify-self` property should apply to block-level elements.

The following tests cover:
* `justify-self: auto | left | right | center | start | end | self-start | self-end | flex-start | flex-end`
* functionality is disabled on `display: float`
* `justify-self: self-[start|end]` respects the direction of the item
* `justify-self: [start|end]` respects the direction of the container
* absolutely positioned items are positioned properly relative to `justify-self` items

See also the following WebKit issue: https://bugs.webkit.org/show_bug.cgi?id=277022

### CODE REVIEW

File: css/css-align/self-alignment/block-justify-self.html [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+  <div style="justify-self: left;">left</div>
+  <div style="justify-self: right;">right</div>
+  <div style="justify-self: start;">start</div>
+  <div style="justify-self: center;">center</div>
+  <div style="justify-self: end;">end</div>
+  <div style="justify-self: self-start;">self-start</div>
+  <div style="justify-self: self-end;">self-end</div>
+  <div style="justify-self: flex-start;">flex-start</div>
+  <div style="justify-self: flex-end;">flex-end</div>
+  <div style="display: float; justify-self: center;">float</div>
```
Comment (fantasai) [inline css/css-align/self-alignment/block-justify-self.html:45]: This should be `float: left` / `float: right`; there's no such thing as `display: float`.

File: css/css-align/self-alignment/block-justify-self.html [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    <div style="justify-self: self-start; direction: rtl">self-start (rtl)</div>
+  <div style="justify-self: self-end; direction: rtl">self-end (rtl)</div>
+  <div style="justify-self: start; direction: rtl">start (rtl)</div>
+  <div style="justify-self: end; direction: rtl">end (rtl)</div>
+  <div style="justify-self: end">end parent
+    <div style="position: absolute;">absolute child</div>
+  </div>
```
Comment (fantasai) [inline css/css-align/self-alignment/block-justify-self.html:52]: This is good, but also test what happens if the aligned `<div>` itself is absolutely positioned.

**Review by fantasai — CHANGES_REQUESTED** [review]
Looks pretty good overall! Two small comments below and one more: make a copy of the entire thing with the container in RTL mode.

(There's a max-width of 600px for the reftest harness IIRC, so you might need to stack those in a second column. The grid trick is quite clever though!)

**Comment by fantasai** [comment]
Ah! Also check the `normal` and `stretch` values to make sure they behave as expected.

**Review by fantasai — APPROVED** [review]
Looks good

File: css/css-align/self-alignment/block-justify-self-ref.html [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    <div style="justify-self: left;">left</div>
+    <div style="justify-self: right;">right</div>
+    <div style="justify-self: start;">start</div>
+    <div style="justify-self: center;">center</div>
+    <div style="justify-self: end;">end</div>
+    <div style="justify-self: self-start;">self-start</div>
+    <div style="justify-self: self-end;">self-end</div>
+    <div style="justify-self: flex-start;">flex-start</div>
+    <div style="justify-self: flex-end;">flex-end</div>
+    <div style="float: left; justify-self: center;">float left</div>
```
Comment (davidsgrogan) [inline css/css-align/self-alignment/block-justify-self-ref.html:43]: grid ignores the `float` property, right?

File: css/css-align/self-alignment/block-justify-self.html [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    <div style="justify-self: left;">left</div>
+    <div style="justify-self: right;">right</div>
+    <div style="justify-self: start;">start</div>
+    <div style="justify-self: center;">center</div>
+    <div style="justify-self: end;">end</div>
+    <div style="justify-self: self-start;">self-start</div>
+    <div style="justify-self: self-end;">self-end</div>
+    <div style="justify-self: flex-start;">flex-start</div>
+    <div style="justify-self: flex-end;">flex-end</div>
+    <div style="float: left; justify-self: center;">float left</div>
```
Comment (davidsgrogan) [inline css/css-align/self-alignment/block-justify-self.html:51]: But the block container doesn't ignore the `float` property. In fact, https://drafts.csswg.org/css-align/#justify-block says about `justify-self`: "This property does not apply to floats."

So when a grid child has both `float` and `justify-self`, `justify-self` wins.

But when a block flow child has both `float` and `justify-self`, `float` wins.

At least this is my interpretation. So I think your -ref file needs something different for these `float`s.

But @fantasai, please let me know if I've misunderstood.

File: css/css-align/self-alignment/block-justify-self.html [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    <div style="justify-self: self-end; direction: rtl">self-end (rtl)</div>
+    <div style="justify-self: start; direction: rtl">start (rtl)</div>
+    <div style="justify-self: end; direction: rtl">end (rtl)</div>
+    <div style="justify-self: end">end parent
+      <div style="position: absolute;">absolute child</div>
+    </div>
+    <div style="justify-self: end; position: absolute;">absolute parent</div>
+  </article>
+
+  <article class="block-container-with-table-items" style="display: block; width: 300px">
```
Comment (davidsgrogan) [inline css/css-align/self-alignment/block-justify-self.html:63]: (Aside: I don't see the purpose of the `block-container-with-table-items` element. Maybe it is leftover from a previous version of these tests and should be deleted? But if it is serving a purpose, could you let me know what it is?)

**Review by davidsgrogan — COMMENTED** [review]
Thanks for these tests, @davidtranhq. Some questions below.

File: css/css-align/self-alignment/block-justify-self.html (hunk 4 above)
Comment (davidsgrogan) [reply css/css-align/self-alignment/block-justify-self.html:51]: @davidtranhq I see that in your commit message you have "functionality is disabled on display: float" which indicates we agree on the intent of the spec. So I suppose you just have to fix the -ref grid item(s) that correspond to these floats.

Sorry I missed that the first time around.

Comment (davidtranhq) [reply css/css-align/self-alignment/block-justify-self.html:51]: Sorry for the late reply, I was working internally on another version of these tests and missed this. I believe I have corrected the issues you were talking about and will make another PR for it :)

**Pull request merged**